adding functionality to build simple circuits from strings

khnikhil opened the pull request:
this code closes #28 as part of the UnitaryHack 2023 event.

Specifically, I added functionality to build simple circuits from strings (for Qiskit) as shown in the following code example:

`circuit_str = "h()^0;CRX(x_0)^01;CRZ(x_1)^10"`
`hierq = Qinit(5)+Qcycle(mapping=Qunitary(circuit_str))`

I show a screenshot comparing this new method with the old method of creating an ansatz function:
<img width="659" alt="Screen Shot 2023-05-26 at 12 39 23 PM" src="https://github.com/matt-lourens/hierarqcal/assets/74874354/cbbf09d4-2210-4baa-95ac-3476b50eb879">

The string must be formatted in the following format:
`"{gate_string}(parameters)^{bits};{gate_string}(parameters)^{bits};...;{gate_string}(parameters)^{bits}"`
where `gate_string` is a Qiskit gate (e.g. one of *{h,x,y,z,cx,cy,cz,rx,ry,rz,crx,cry,crz,s,toffoli}*) and can be written in uppercase, lowercase, or a mix of cases; `parameters` is a string of format *{x0, x1, x2, ..., xn}*; bits is the qubit/set of qubits on which the gate is applied.



**NB: Each gate instruction MUST be separated by a semicolon**
**NB: you do not have to specify `arity` and `n_symbols`, as these are extracted automatically from the string in the Qunitary `__init__()` function.**

As specified in issue #28, I added code in `core.py` in the `Qunitary.__init__()` function to check if the input function is a string, and if so, automatically calculate and set the `arity` and `n_symbols` attributes. Then, I modified the `get_circuit_qiskit()` function to convert the string to a Python function, using a helper function in `utils.py`.

@matt-lourens As this is my first time contributing to an open-source project, please let me know if I'm missing anything! thanks so much :)

matt-lourens commented:
Hey @khnikhil, this is awesome!

I will still need to go through some of the details, but from an overall glance and some tests, this is what I was looking for.  While I go through the details of the PR, something that you can add so long is the following:

So providing a hierarchical object as a mapping doesn't work anymore (check the **Sub motif as a mapping for a motif** section in the quickstarti.ipynb). Don't worry though, this is an involved issue because I have this Qhierarchy.get_unitary_function in core.py that converts a Qhierarchy object into a Qunitary.function (this is to build up motifs in terms smaller motifs).  I have the solution here, just change the following:

On line 1158 in core.py update the `get_unitary_function` as follows:
```python
    def get_unitary_function(self, **kwargs):
        """
        Convert the Qhierarchy into a function that can be called.
        """

        def unitary_function(bits, symbols=None, **kwargs):
            self.update_Q(bits)
            if not (symbols is None):
                self.set_symbols(symbols)
            return_object = None
            for layer in self:
                for unitary in layer.edge_mapping:
                    if isinstance(unitary.function, str):
                        get_circuit_from_string = kwargs.get("get_circuit_from_string", None)
                        unitary = get_circuit_from_string(unitary)
                    return_object = unitary.function(
                        unitary.edge, unitary.symbols, **kwargs
                    )
            return return_object

        return unitary_function
```
Then in your new qiskit_helper.get_circuit_qiskit function allow it to recieve \*\*kwargs: on **line 97**
```python
   # breaking down the qunitary.function string into a list of gate instructions
    instruction_list, unique_bits, _ = get_circ_info_from_string(qunitary.function)


    # building a function from the list of gate instructions
    def circuit_fn(bits, symbols=None, circuit=None, **kwargs):
        qubits = [QuantumRegister(1, bits[i]) for i in range(qunitary.arity)]
```
Then on **line 73** in qiskit_helper.get_circuit_qiskit send your function to the unitary.function call:

```python
            circuit = unitary.function(
                bits=unitary.edge, symbols=unitary.symbols, circuit=circuit, get_circuit_from_string=get_circuit_from_string
            )
```
 
The idea here is to send along the backends specific string to circuit conversion function to the abstract get_unitary_function in core.py. Since we want to allow something as follows:
```python
m1 = Qinit(3) + Qpermute(mapping=Qunitary("crx(x_0)^01"))
m2 = Qinit(6) + Qcycle(mapping=m1)
m3 = Qinit(18) + Qcycle(mapping=m2)
```
We need to make sure whatever that string conversion function is, needs to be accessible on the lowest level of the compute graph. So the only requirement is that if you provide a function, just specify **kwargs which allows us to send around backend specific functions like string to circuit conversions. 

One final small change in regards to this is just the default circuits in qiskit_circuits.py, just give each of them a **kwargs parameter:
```python
def U2(bits, symbols=None, circuit=None, **kwargs):
def U3(bits, symbols=None, circuit=None, **kwargs):
def V2(bits, symbols=None, circuit=None, **kwargs):
def V4(bits, symbols=None, circuit=None, **kwargs):
```

As far as I can tell,  with these changes nothing breaks and works as expected. 

Tomorrow I will go through the details of the PR, but any changes from here will be small stuff, great job!

khnikhil commented:
Hi @matt-lourens,

Thanks so much for the quick comments! I made the changes you pointed out, and everything seems to be working fine.

Just to make sure I understand, the idea is that we want to pass on the `get_circuit_from_string` method through kwargs so that it can be used higher up the stack when we're converting a `Qhierarchy` object into a single function?

I'm a bit confused by what you mean by the following: 

> So providing a hierarchical object as a mapping doesn't work anymore (check the **Sub motif as a mapping for a motif** section in the quickstart.ipynb). 

since in `quickstart.ipynb` it seems like the hierarchical object `m1` is still being passed as a mapping to `hierq`:

<img width="972" alt="Screen Shot 2023-05-27 at 11 24 08 AM" src="https://github.com/matt-lourens/hierarqcal/assets/74874354/0a76816d-0dbf-45f2-8b74-9970a80dc323">

Thanks so much for the comments, and please let me know if there are any other fixes that should be made!

matt-lourens commented on hierarqcal/core.py:
```diff
@@ -47,8 +47,47 @@ def __init__(self, function=None, n_symbols=0, arity=2, symbols=None):
+                else:
+                    param_list = params_string.split(',')
+                    # Remove any leading or trailing whitespaces from each substring
+                    param_list = [param_entry.strip() for param_entry in param_list]
```
Minor thing: Can we change this so that param_list only adds param_entries that is not in unique_params. i.e. do the if check within this list comprehension and then extend unique_params on the next line if param_list is not empty. This way we don't have to loop twice through param_list

khnikhil replied:
yes, done

matt-lourens commented on hierarqcal/utils.py:
```diff
@@ -205,4 +205,96 @@ def plot_motif(
+#     return figs
+
+
+def get_circ_info_from_string(input_str):
```
The code is very robust and working well! The logic is sound.

I have a suggestion around the organisation of the functions:
I think this function should be part of `Qunitary`, both this code and Qunitary's intialization calculates `unique_params` and `unique_bits`. Since getting the `substr_list` is backend independant it can form part of Qunitary and run when initialized. During initialization store substr_list as an attribute of Qunitary (which by default should be None unless a string was provided). With substr_list as an attribute for Qunitary, I think a good name would be circuit_instructions and we should change it's type to a namedTuple (then we can reference the instructions by key) i.e. circuit_instructions.gate_name, circuit_instructions.symbol_info, circuit_instructions.sub_bits

So what I'm thinking about looks something along the lines:
```python
# core.py
from collections import namedtuple
# ...
CircuitInstructions = namedtuple('CircuitInstructions', ['gate_name', 'symbol_info', 'sub_bits'])
# ...
class Qunitary():
     def __init__
          self.circuit_instructions = None
          if isinstance(function, str):
                circuit_instructions, unique_params, unique_bits = self.get_circuit_from_string(function)
                # Do the self.attribute = stuff 


```
Then the get_circuit_from_string in qiskit helper stays the same, only it gets circuit instructions directly from the qunitary object sent to it, i.e. `qunitary.circuit_instructions.gate_name` etc

matt-lourens replied:
Looking at this again, I realize it should rather be:
```python
CircuitInstruction= namedtuple('CircuitInstruction', ['gate_name', 'symbol_info', 'sub_bits'])
```
So that every circuit_instruction is a named tuple, then the Qunitary has the attribute circuit_instructions which is a list `[CircuitInstruction(...), CircuitInstruction(...)]`

khnikhil replied:
ah I see, this is nice! implemented

matt-lourens commented on hierarqcal/qiskit/qiskit_helper.py:
```diff
@@ -65,10 +66,57 @@ def get_circuit_qiskit(hierq, symbols=None, barriers=True):
         for unitary in layer.edge_mapping:
+
+            if isinstance(unitary.function, str):
+                unitary = get_circuit_from_string(unitary)
```
Can you add a comment here that says we're updating the qunitary object's function attribute to a qiskit function based on the provided string

khnikhil replied:
yes, thanks for the note!

matt-lourens reviewed (commented):
@khnikhil This is great! I've left a few comments, but it's mostly about the way we organize the code. In terms of functionality, everything is perfect!

matt-lourens commented:
> Hi @matt-lourens,
> 
> Thanks so much for the quick comments! I made the changes you pointed out, and everything seems to be working fine.
> 
> Just to make sure I understand, the idea is that we want to pass on the `get_circuit_from_string` method through kwargs so that it can be used higher up the stack when we're converting a `Qhierarchy` object into a single function?
> 
> I'm a bit confused by what you mean by the following:
> 
> > So providing a hierarchical object as a mapping doesn't work anymore (check the **Sub motif as a mapping for a motif** section in the quickstart.ipynb).
> 
> since in `quickstart.ipynb` it seems like the hierarchical object `m1` is still being passed as a mapping to `hierq`:
> <img alt="Screen Shot 2023-05-27 at 11 24 08 AM" width="972" src="https://user-images.githubusercontent.com/74874354/241422081-0a76816d-0dbf-45f2-8b74-9970a80dc323.png">
> 
> Thanks so much for the comments, and please let me know if there are any other fixes that should be made!

So in regards to this, on the high level (interfacing) you pass the Qhierarchy object as a mapping, but on the low level when a Qmotif is intialised and it's mapping is a Qhierarchy, then the mapping gets updated to a function by calling `get_unitary_function()`:
```python
# core.Qmotif
        if self.is_default_mapping:
            self.arity = 2
        else:
            if isinstance(self.mapping, Qhierarchy):
                # If mapping was specified as sub-hierarchy, convert it to a qunitary
                new_mapping = Qunitary(
                    function=None,
                    n_symbols=self.mapping.n_symbols,
                    arity=len(self.mapping.tail.Q),
                )
                new_mapping.function = self.mapping.get_unitary_function()
                self.mapping = new_mapping
            self.arity = self.mapping.arity
```
Before my suggested change when this self.mapping  was a Qhierarchy which had a `Qmotif` with a mapping as string, the .get_unitary_function() bombs out because that `Qmotif` never got it's qiskit converted function to execute. We therefore have to pass that conversion function all the way down (no matter how many layers)during execution, so that the function correctly gets converted to the function we want. It's a bit hard to explain, but it's an interesting problem and solution, if you're interested undo the changes  I suggested and debug through the sub_hierq and sub sub hierq examples of the notebook. Then it's a bit more clear what goes wrong :D

khnikhil commented:
> @khnikhil This is great! I've left a few comments, but it's mostly about the way we organize the code. In terms of functionality, everything is perfect!

hi @matt-lourens! thanks for the comments - I implemented everything in the new commits. in addition to moving the `get_circ_info_from_string` function inside `Qunitary`, I changed the name of the function I added to `qiskit_helper.py` to `get_qiskit_circuit_from_instructions`, which should make the code easier to understand in addition to the extra comments. thanks so much for the comments and support again, this was really fun :)

khnikhil commented:
hi @matt-lourens! I just realized I was mislabelling `unique_bits` and `unique_params` when calling them in `Qunitary.__init__()` which was leading to some bugs, so I went back in and fixed it. Is there anything else I need to do before it can be merged?

matt-lourens reviewed (approved):
@khnikhil  Nice! all is working on my side, and the new organization is great. I'm going to merge and assign the issue to you (for the bounty) :D
